python/cuml/experimental/explainer/kernel_shap.pyx

```diff
@@ -0,0 +1,480 @@
+            self.feature_names = data.columns.to_list()
+        else:
+            self.feature_names = [None for _ in range(len(data))]
+
+        # seeing how many exact samples from the powerset we can enumerate
+        # todo: optimization for larger sizes by generating diagonal
+        # and gpu lexicographical-binary numbers generation
+        cur_nsamples = self.M
+        r = 1
+        while cur_nsamples < self.nsamples:
+            if has_scipy():
+                from scipy.special import binom
+                cur_nsamples += int(binom(self.M, r))
+            else:
+                cur_nsamples += int(binomCoef(self.M, r))
+            r += 1
+
+        # see if we need to have randomly sampled entries in our mask
+        # and combinations matrices
+        self.nsamples_random = max(self.nsamples - cur_nsamples, 0)
+
+        # using numpy powerset and calculations for initial version
+        # cost is incurred only once, and generally we only generate
+        # very few samples if M is big.
+        mat, weight = powerset(self.M, r, self.nsamples, dtype=self.dtype)
+        weight /= np.sum(weight)
+
+        self.mask, *_ = input_to_cuml_array(mat)
+        self.nsamples_exact = len(self.mask)
+
+        self.weights = cp.empty(self.nsamples, dtype=self.dtype)
+        self.weights[:self.nsamples_exact] = cp.array(weight)
+
+        self.synth_data = None
+
+        self.expected_value = self.link_fn(cp.mean(model(self.background)))
+
+        self.random_state = random_state
+
+    def explain(self,
+                X,
+                nsamples=None,
+                l1_reg='auto'):
+        shap_values = cp.zeros((1, self.M), dtype=self.dtype)
+
+        # allocating combinations array once for multiple explanations
+        if self.synth_data is None:
+            self.synth_data = CumlArray.zeros(
+                shape=(self.N * self.nsamples, self.M),
+                dtype=np.float32,
+                order='C'
+            )
+
+        idx = 0
+        for x in X:
+            shap_values[idx, :-1] = self._explain_single_observation(x, l1_reg)
+
+            # we need to add the last value since we removed one variable
+            res = (self.link_fn(self.fx) -
+                   self.link_fn(self.expected_value)) - cp.sum(shap_values)
+
+            shap_values[idx, -1] = res[0]
+            idx += 1
+
+        if isinstance(X, np.ndarray):
+            out_type = 'numpy'
+        else:
+            out_type = 'cupy'
+        return input_to_cuml_array(shap_values).to_output(out_type)
+
+    def _explain_single_observation(self,
+                                    row,
+                                    l1_reg):
+
+        # np choice of weights - for samples if needed
+        # choice algorithm can be optimized for large dimensions
+
+        self.fx, *_ = input_to_cuml_array(self.model(row.reshape(1, self.M)))
+
+        if self.nsamples_random > 0:
```

Comment:
This explain function has also gotten pretty long and may be worth breaking up a little and unit testing separately

Reply:
Also, there is a TON of code without any comments. Could use some more info here.